examples/auth.rs

```diff
@@ -1,45 +1,279 @@
+}
 
-fn perform_auth_request() -> duchess::GlobalResult<String> {
-    Jvm::with(|jvm| {
-        let params = HashMap::new().execute(jvm)?;
-        let values = ArrayList::new().execute(jvm)?;
-        values.add("first-value").execute(jvm)?;
-        values.add("second-value").execute(jvm)?;
-        params.put("first-param", &values).execute(jvm)?;
+#[derive(Debug)]
+pub enum AuthenticateError {
+    Unathenticated(String),
+    InvalidSecurityToken,
+    InvalidSignature,
+    InternalError(String),
+}
 
-        let http_request =
-            HttpRequest::new("POST", "/", [1i8, 2, 3].as_slice(), &params).execute(jvm)?;
+#[derive(Debug)]
+pub struct AuthorizeRequest {
+    pub resource: String,
+    pub action: String,
+    pub context: HashMap<String, String>,
+}
 
-        Ok(http_request.to_string().assert_not_null().into_rust(jvm)?)
-    })
+#[derive(Debug)]
+pub enum AuthorizeError {
+    Denied(String),
+    InternalError(String),
 }
 
-fn main() -> duchess::GlobalResult<()> {
-    let s = perform_auth_request()?;
-    println!("{s}");
-    Ok(())
+impl HttpAuth {
+    pub fn new() -> duchess::GlobalResult<Self> {
+        let auth = Jvm::with(|jvm| {
+            let auth = java_auth::HttpAuth::new().execute(jvm)?;
+            Ok(jvm.global(&*auth))
+        })?;
+        Ok(Self(auth))
+    }
+
+    pub fn authenticate(&self, request: &HttpRequest) -> Result<Authenticated, AuthenticateError> {
+        Jvm::with(|jvm| {
+            let request = request.into_java(jvm)?;
+            match self.0.authenticate(&request).assert_not_null().execute(jvm) {
+                Ok(auth) => Ok(Ok(Authenticated::from_java(jvm, auth)?)),
+                Err(duchess::Error::Thrown(exception)) => {
+                    // XX: is this kind of type switching better handled by a macro?
+                    Ok(Err(
+                        // XX: why can't we infer the <Throwable, ? 
+                        if let Ok(x) = exception.try_downcast::<Throwable, java_auth::AuthenticationExceptionUnauthenticated>().execute(jvm)? {
+                            let message = x.user_message().assert_not_null().into_rust(jvm)?;
+                            AuthenticateError::Unathenticated(message)
+                        // XX: should we add a .is_instance() alias for try_downcast().is_ok()?
+                        } else if exception.try_downcast::<Throwable, java_auth::AuthenticationExceptionInvalidSecurityToken>().execute(jvm)?.is_ok() {
+                            AuthenticateError::InvalidSecurityToken
+                        } else if exception.try_downcast::<Throwable, java_auth::AuthenticationExceptionInvalidSignature>().execute(jvm)?.is_ok() {
+                            AuthenticateError::InvalidSignature
+                        } else {
+                            let message = exception.get_message().assert_not_null().into_rust(jvm)?;
+                            AuthenticateError::InternalError(message)
+                        }
```

Comment:
Wild idea: could we add a macro'd enum of all downcastable types listed in the `java_package!` invocation? Something like

```rust
duchess::java_package! {
    package auth;
    
    // ...
    #[downcasting]
    class AuthenticationException { * }
    class Unauthenticated { * }
    class InvalidSecurityToken { * }
    class InvalidSignature { * }
}
```

which would then let this code look like

```rust
match exception.downcasting(jvm)? {
    AuthenticationExceptionDowncast::Unauthenticated(x) => { .. }
    AuthenticationExceptionDowncast::InvalidSecurityToken(_x) => { .. }
    AuthenticationExceptionDowncast::Else(x) => { .. }
}
```

Reply:
That's a neat idea. I think we could do *something* like that yes. I like the idea of trying to make it possible to make authoring the Java code a bit more "rust-y".